analytics/agma/agma_module.go

```diff
@@ -135,18 +139,15 @@ func (l *AgmaLogger) flush() {
 	l.buffer.Write([]byte("]"))
 
 	payload := make([]byte, l.buffer.Len())
-	_, err := l.buffer.Read(payload)
-	if err != nil {
-		l.reset()
-		l.mux.Unlock()
+	numBytesCopied, err := l.read(payload, l.buffer)
+	l.reset()
+	l.mux.Unlock()
+	if err != nil || numBytesCopied == 0 {
 		glog.Warning("[AgmaAnalytics] fail to copy the buffer")
 		return
 	}
 
 	go l.sender(payload)
-
-	l.reset()
-	l.mux.Unlock()
```

Comment:
Given that `payload` is a copy, `l.reset()` and `l.mux.Unlock()` should be moved upstream so the memory is unlocked and available for other processes in case `go l.sender(payload)` gets run before the remainder of `flush()`.